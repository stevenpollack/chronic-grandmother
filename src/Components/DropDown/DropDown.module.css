.container {
    display: flex;
    flex-direction: column;
    position: relative;
    max-width: 400px;

    font-family: 'Nunito Sans';
    color: #0b1a3f;
}

.dropdown {
    display: flex;
    justify-content: flex-start;
    align-items: center;
    max-width: 400px;
    width: 400px;
    height: 48px;
    background: #fff;
    border: 1px solid #e5e5e5;
    border-radius: 4px;
}

.dropdownText {
    font-size: 16px;
}

.dropdownToggle {
    margin-left: auto;
}
.toggleArrow {
    transform-box: fill-box;
    transform-origin: center;
    transform: rotate(0deg);
    transition: all 0.2s;
}
.toggleArrow.toggled {
    transform: rotate(180deg);
    transition: all 0.2s;
}

.button {
    display: flex;
    flex-direction: row;
}

.menu {
    position: absolute;

    list-style-type: none;
    margin: 5px 0;
    padding: 0;

    background: #fff;
    border-radius: 5px;
    box-shadow: 0 5px 10px rgba(0, 0, 0, 0.15);
    width: 100%;
    top: 70px;
    max-height: 150px;
    overflow-y: auto;
    z-index: 1000;
}
.menu > li {
    margin: 0;

    background-color: white;
}
.menu > li:hover {
    background: rgba(85, 127, 233, 0.1);
    color: #457ced;
    cursor: pointer;
}
.menu > li > button {
    width: 100%;
    height: 100%;
    text-align: left;

    background: none;
    color: inherit;
    border: none;
    padding: 5px;
    margin: 0;
    font: inherit;
    cursor: pointer;
    color: #0b1a3f;
    font-size: 16px;
}
