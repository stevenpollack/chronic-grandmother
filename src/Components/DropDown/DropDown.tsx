import { useState } from 'react';

import classes from './DropDown.module.css';

const DropDown = (props) => {
    const [open, setOpen] = useState(false);

    const handleOpen = () => {
        setOpen(!open);
    };

    const handleSelect = (key) => {
        props.setSelected(key);
        setOpen(false);
    };

    return (
        <div className={`${classes.container} ${props.className}`} style={props.style}>
            {props.label && <span>{props.label}</span>}
            <button onClick={handleOpen} className={classes.dropdown}>
                {props.leftIcon}
                <span className={classes.dropdownText}>{props.selected}</span>

                <div className={classes.dropdownToggle}>
                    <svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                        <path
                            d="M16.2 5.3894H7.8C6.2536 5.3894 5 6.64301 5 8.1894V16.5894C5 18.1358 6.2536 19.3894 7.8 19.3894H16.2C17.7464 19.3894 19 18.1358 19 16.5894V8.1894C19 6.64301 17.7464 5.3894 16.2 5.3894Z"
                            stroke="#E5E5E5"
                            strokeWidth="1.03427"
                        />

                        <path
                            className={`${classes.toggleArrow} ${open ? classes.toggled : ''}`}
                            fillRule="evenodd"
                            clipRule="evenodd"
                            d="M9.20711 11.3894C8.76165 11.3894 8.53857 11.928 8.85355 12.243L11.6464 15.0359C11.8417 15.2311 12.1583 15.2311 12.3536 15.0358L15.1464 12.243C15.4614 11.928 15.2383 11.3894 14.7929 11.3894H9.20711Z"
                            fill="#6D7587"
                        />
                    </svg>
                </div>
            </button>
            {open ? (
                <ul className={classes.menu}>
                    {props.options.map(({ option, key, icon }) => (
                        <li className={classes['menu-item']}>
                            <button className={classes.button} onClick={() => handleSelect(key)}>
                                {icon}
                                {option}
                            </button>
                        </li>
                    ))}
                </ul>
            ) : null}
        </div>
    );
};

// DropDown.propTypes = {
//     selected: PropTypes.string,
//     setSelected: PropTypes.func,
//     label: PropTypes.string,
//     options: PropTypes.array,
//     leftIcon: PropTypes.object,
//     className: PropTypes.string,
//     style: PropTypes.object,
// };

export default DropDown;
