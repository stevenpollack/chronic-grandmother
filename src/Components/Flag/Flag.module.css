.flag {
  margin-right: 10px;
  width: 26px;
  height: 20px;
  margin-left: 5px;
}
