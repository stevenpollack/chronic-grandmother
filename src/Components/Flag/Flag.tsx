import classes from "./Flag.module.css";

export function Flag({ code }: { code: string }) {
  return (
    <img
      alt={code || ""}
      src={`/img/flags/${code || ""}.svg`}
      width="20px"
      className={classes.flag}
      data-testid={`${code}-flag-icon`}
    />
  );
}
