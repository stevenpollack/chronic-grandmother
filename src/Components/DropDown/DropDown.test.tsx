import DropDown, { DropDownProps } from "./DropDown";
import userEvent from "@testing-library/user-event";
import { render, waitFor } from "@testing-library/react";
import { Flag } from "../Flag/Flag";

const OPTIONS: DropDownProps["options"] = [
  { option: "AUD", key: "AU", icon: <Flag code="AU" /> },
  { option: "USD", key: "US", icon: <Flag code="US" /> },
  { option: "GBP", key: "GB", icon: <Flag code="GB" /> },
  { option: "JPY", key: "JP", icon: <Flag code="JP" /> },
];

describe("DropDown", () => {
  const defaultProps: DropDownProps = {
    selected: OPTIONS[0].option,
    setSelected: vi.fn(),
    label: "From",
    options: OPTIONS,
    leftIcon: OPTIONS[0].icon,
    style: {},
  };

  function renderDropDown(props?: DropDownProps) {
    const mergedProps = { ...defaultProps, ...props };

    return {
      user: userEvent.setup(),
      ...render(<DropDown {...mergedProps} />),
    };
  }

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders correctly with provided props", () => {
    const { getByTestId, asFragment } = renderDropDown();

    expect(getByTestId(`${defaultProps.label}-value`)).toHaveTextContent(
      defaultProps.selected
    );
    expect(getByTestId(`${OPTIONS[0].key}-flag-icon`)).toBeVisible();
    expect(getByTestId("dropdown-button")).toBeVisible();
    expect(getByTestId("dropdown-button")).toBeEnabled();

    // match snapshot
    expect(asFragment()).toMatchSnapshot();
  });

  it("presents options when button is clicked", async () => {
    const { user, getByTestId, getByText } = renderDropDown();

    await user.click(getByTestId("dropdown-button"));

    await waitFor(() => {
      // we don't need to test the first option as it is already selected
      OPTIONS.slice(1).forEach((option) => {
        expect(getByText(option.option)).toBeInTheDocument();
      });
    });
  });

  it("calls setSelected when option is clicked", async () => {
    const { user, getByTestId, getByText } = renderDropDown();

    await user.click(getByTestId("dropdown-button"));

    await waitFor(async () => {
      const usdOption = getByText(OPTIONS[1].option);
      await user.click(usdOption);
    });

    expect(defaultProps.setSelected).toHaveBeenCalledWith(OPTIONS[1].key);
  });

  it("closes dropdown when option is selected", async () => {
    const { user, getByTestId, getByText, queryByRole, getByRole } =
      renderDropDown();

    await user.click(getByTestId("dropdown-button"));

    await waitFor(async () => {
      expect(getByRole("listbox")).toBeInTheDocument();
      const usdOption = getByText(OPTIONS[1].option);
      await user.click(usdOption);
    });

    await waitFor(() => {
      expect(queryByRole("listbox")).toBeNull();
    });
  });

  it("closes dropdown when clicking outside", async () => {
    const { user, getByTestId, queryByRole, getByRole, container } =
      renderDropDown();

    await user.click(getByTestId("dropdown-button"));

    await waitFor(() => {
      expect(getByRole("listbox")).toBeInTheDocument();
    });

    // Click outside the dropdown
    await user.click(container);

    await waitFor(() => {
      expect(queryByRole("listbox")).toBeNull();
    });
  });

  it("toggles dropdown state when clicking button multiple times", async () => {
    const { user, getByTestId, queryByRole, getByRole, container } =
      renderDropDown();

    // Open dropdown
    await user.click(getByTestId("dropdown-button"));
    await waitFor(() => {
      expect(getByRole("listbox")).toBeInTheDocument();
    });

    // Close dropdown
    await user.click(getByTestId("dropdown-button"));
    await waitFor(() => {
      expect(queryByRole("listbox")).toBeNull();
    });
  });
});
