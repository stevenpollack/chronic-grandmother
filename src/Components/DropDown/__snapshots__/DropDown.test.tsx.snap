// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropDown > renders correctly with provided props 1`] = `
<DocumentFragment>
  <div
    class="_container_5d938f undefined"
  >
    <label
      class="_label_5d938f"
      id="From-label"
    >
      From
    </label>
    <button
      aria-describedby="From-value"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="From-label"
      class="_dropdown_5d938f"
      data-testid="dropdown-button"
    >
      <img
        alt="AU"
        class="_flag_479884"
        data-testid="AU-flag-icon"
        src="/img/flags/AU.svg"
        width="20px"
      />
      <span
        class="_dropdownText_5d938f"
        data-testid="From-value"
      >
        AUD
      </span>
      <div
        aria-hidden="true"
        class="_dropdownToggle_5d938f"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16.2 5.3894H7.8C6.2536 5.3894 5 6.64301 5 8.1894V16.5894C5 18.1358 6.2536 19.3894 7.8 19.3894H16.2C17.7464 19.3894 19 18.1358 19 16.5894V8.1894C19 6.64301 17.7464 5.3894 16.2 5.3894Z"
            stroke="#E5E5E5"
            stroke-width="1.03427"
          />
          <path
            class="_toggleArrow_5d938f "
            clip-rule="evenodd"
            d="M9.20711 11.3894C8.76165 11.3894 8.53857 11.928 8.85355 12.243L11.6464 15.0359C11.8417 15.2311 12.1583 15.2311 12.3536 15.0358L15.1464 12.243C15.4614 11.928 15.2383 11.3894 14.7929 11.3894H9.20711Z"
            fill="#6D7587"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </button>
  </div>
</DocumentFragment>
`;
